torchx/lib/torchx/backend.ex

```diff
@@ -981,66 +987,78 @@ defmodule Torchx.Backend do
 
   @impl true
   def window_max(out, tensor, window_dims_tuple, opts) do
-    if tuple_size(window_dims_tuple) > 3 do
-      raise ArgumentError, "cannot have more than 3 dimensions for the window"
+    window_op(
+      out,
+      tensor,
+      window_dims_tuple,
+      opts,
+      &Torchx.amax(&1, &2, false)
+    )
+  end
+
+  @impl true
+  def window_min(out, tensor, window_dims_tuple, opts) do
+    window_op(
+      out,
+      tensor,
+      window_dims_tuple,
+      opts,
+      &Torchx.amin(&1, &2, false)
+    )
+  end
+
+  @impl true
+  def window_sum(out, tensor, window_dims_tuple, opts) do
+    window_op(out, tensor, window_dims_tuple, opts, &Torchx.sum(&1, &2, false))
+  end
+
+  @impl true
+  def window_product(out, tensor, window_dims_tuple, opts) do
+    window_op(out, tensor, window_dims_tuple, opts, fn tensor, axes ->
+      aggregate_over_axes(tensor, axes, false, &Torchx.product/3)
+    end)
+  end
+
+  def window_op(out, tensor, window_dims_tuple, opts, reduce_fun)
+      when is_function(reduce_fun, 2) do
+    if opts[:window_dilations] != List.duplicate(1, tuple_size(tensor.shape)) do
+      raise ArgumentError, "window_dilations unsupported"
     end
 
     if Enum.any?(opts[:padding], fn conf -> conf |> Tuple.to_list() |> Enum.any?(&(&1 != 0)) end) do
       raise ArgumentError, "padding unsupported"
     end
 
     strides = opts[:strides]
-    window_dilations = opts[:window_dilations]
-
-    strides = List.duplicate(1, 3 - length(strides)) ++ strides
-    window_dilations = List.duplicate(1, 3 - length(window_dilations)) ++ window_dilations
-
-    window_dims =
-      List.duplicate(1, 3 - tuple_size(window_dims_tuple)) ++ Tuple.to_list(window_dims_tuple)
-
-    pool_shape =
-      List.to_tuple(
-        List.duplicate(1, 4 - tuple_size(tensor.shape)) ++ Tuple.to_list(tensor.shape)
-      )
-
-    # TO-DO: implement actual padding
-    padding = List.duplicate(0, 3)
 
     intermediate_type =
       tensor.type
       |> Nx.Type.to_floating()
       |> to_torch_type()
 
-    tensor
-    |> from_nx()
-    |> Torchx.reshape(pool_shape)
-    |> Torchx.to_type(intermediate_type)
-    |> Torchx.max_pool_3d(window_dims, strides, padding, window_dilations)
-    |> Torchx.reshape(out.shape)
-    |> Torchx.to_type(to_torch_type(out.type))
-    |> to_nx(out)
-  end
-
-  @impl true
-  def window_min(out, tensor, window_dims_tuple, opts) do
-    # libtorch does not expose min_pool3d or similar
-    # Therefore, we use the fact that for any set of negative numbers,
-    # the maximum value is also the one with the smallest absolute value.
-
-    # For instance abs(min([1, 2, 3, 4])) == abs(max([-1, -2, -3, -4])).
+    # TO-DO: support padding
```

Comment:
Padding is not supported in this PR because for some reason Nx.Constants.max_finite and min_finite aren't playing well with libtorch

Reply:
For some reason Torchx was behaving a lot differently if we passed an Nx.tensor instead of a number for the padding constant. This has been fixed.